src/naive_bayes/mod.rs

```diff
@@ -0,0 +1,239 @@
+use crate::error::Failed;
+use crate::linalg::BaseVector;
+use crate::linalg::Matrix;
+use crate::math::num::RealNumber;
+use std::marker::PhantomData;
+
+/// Distribution used in the Naive Bayes classifier.
+pub trait NBDistribution<T: RealNumber, M: Matrix<T>> {
+    /// Prior of class k
+    fn prior(&self, k: T) -> T;
+
+    /// Conditional probability of feature j given class k
+    fn conditional_probability(&self, k: T, j: &M::RowVector) -> T;
+
+    /// Possible classes of the distribution
+    fn classes(&self) -> &Vec<T>;
+}
+
+/// Base struct for the Naive Bayes classifier.
+pub struct BaseNaiveBayes<T: RealNumber, M: Matrix<T>, D: NBDistribution<T, M>> {
+    distribution: D,
+    _phantom_t: PhantomData<T>,
+    _phantom_m: PhantomData<M>,
+}
+
+impl<T: RealNumber, M: Matrix<T>, D: NBDistribution<T, M>> BaseNaiveBayes<T, M, D> {
+    /// Fits NB classifier to a given NBdistribution.
+    /// * `distribution` - NBDistribution of the training data
+    pub fn fit(distribution: D) -> Result<Self, Failed> {
+        Ok(Self {
+            distribution,
+            _phantom_t: PhantomData,
+            _phantom_m: PhantomData,
+        })
+    }
+
+    /// Estimates the class labels for the provided data.
+    /// * `x` - data of shape NxM where N is number of data points to estimate and M is number of features.
+    /// Returns a vector of size N with class estimates.
+    pub fn predict(&self, x: &M) -> Result<M::RowVector, Failed> {
+        let y_classes = self.distribution.classes();
+        let (rows, _) = x.shape();
+        let predictions = (0..rows)
+            .map(|row_index| {
+                let row = x.get_row(row_index);
+                let (prediction, _probability) = y_classes
+                    .iter()
+                    .map(|class| {
+                        (
+                            class,
+                            self.distribution.conditional_probability(*class, &row)
+                                * self.distribution.prior(*class),
+                        )
+                    })
+                    .max_by(|(_, p1), (_, p2)| p1.partial_cmp(p2).unwrap())
+                    .unwrap();
+                *prediction
+            })
+            .collect::<Vec<T>>();
+        let mut y_hat = M::RowVector::zeros(rows);
+        for (i, prediction) in predictions.iter().enumerate().take(rows) {
+            y_hat.set(i, *prediction);
+        }
+        Ok(y_hat)
+    }
+}
+
+/// Naive Bayes classifier for categorical features
+pub struct CategoricalNB<T: RealNumber> {
+    class_labels: Vec<T>,
+    class_probabilities: Vec<T>,
+    coef: Vec<Vec<Vec<T>>>,
+    feature_categories: Vec<Vec<T>>,
+}
+
+impl<T: RealNumber, M: Matrix<T>> NBDistribution<T, M> for CategoricalNB<T> {
+    fn prior(&self, k: T) -> T {
+        match self.class_labels.iter().position(|&t| t == k) {
```

Comment:
I feel that there should be a way to avoid searching for a position in array where t == k. While we can assume that number of classes is small, usually < 100 it is better to avoid inefficiencies if possible.  
One way to make sure we do not search array every time when we need to get a prior is to store all distinct class labels in a vector and use position of the class in a vector as an ID for this class. This way you know that you class labels starts from 0 and go up to k, where k is number of classes. In this case your priors will be stored in a vector of length k with position of each element corresponding to a prior of class your class.
To go back to "real" class values all you need is to replace class IDs with real class value from the stored vector.

Reply:
Do you mean, prior should receive as an input an index instead of the actual value of the label?

Reply:
yes for this method and for `conditional_probability`.

Reply:
implemented in https://github.com/smartcorelib/smartcore/pull/15/commits/456249bdb8ce466edbf5747b21d014cfecccfb38